evm/src/arithmetic/byte.rs

```diff
@@ -0,0 +1,502 @@
+//! given X as an array of 16-bit limbs; write X := \sum_{i=0}^15 Y_i
+//! 2^{16i} where 0 <= Y_i < 2^16.
+//!
+//! The technique (hat tip to Jacqui for the idea) is to store a tree
+//! of limbs of X that are selected according to the bits in I.  The
+//! main observation is that each bit `bi` halves the number of
+//! candidate bytes that we might return: If b4 is 0, then I < 16 and
+//! the possible bytes are in the top half of X: Y_8,..,Y_15
+//! (corresponding to bytes X_16,..,X_31), and if b4 is 1 then I >= 16
+//! and the possible bytes are the bottom half of X: Y_0,..,Y_7
+//! (corresponding to bytes X_0,..,X_15).
+//!
+//! Let Z_0,..,Z_7 be the bytes selected in the first step. Then, in
+//! the next step, if b3 is 0, we select Z_4,..,Z_7 and if it's 1 we
+//! select Z_0,..,Z_3. Together, b4 and b3 divide the bytes of X into
+//! 4 equal-sized chunks of 4 limbs, and the byte we're after will be
+//! among the limbs 4 selected limbs.
+//!
+//! Repeating for b2 and b1, we reduce to a single 16-bit limb
+//! L=x+y*256; the desired byte will be x if b0 is 1 and y if b0
+//! is 0.
+//!
+//! -*-
+//!
+//! To prove that the bytes x and y are in the range [0, 2^8) (rather
+//! than [0, 2^16), which is all the range-checker guarantees) we do
+//! the following (hat tip to Jacqui for this trick too): Instead of
+//! storing x and y, we store w = 256 * x and y. Then, to verify that
+//! x, y < 256 and the last limb L = x + y * 256, we check that
+//! L = w / 256 + y * 256.
+//!
+//! The proof of why verifying that L = w / 256 + y * 256
+//! suffices is as follows:
+//!
+//! 1. The given L, w and y are range-checked to be less than 2^16.
+//! 2. y * 256 ∈ {0, 256, 512, ..., 2^24 - 512, 2^24 - 256}
+//! 3. w / 256 = L - y * 256 ∈ {-2^24 + 256, -2^24 + 257, ..., 2^16 - 2, 2^16 - 1}
+//! 4. By inspection, for w < 2^16, if w / 256 < 2^16 or
+//!    w / 256 >= P - 2^24 + 256 (i.e. if w / 256 falls in the range
+//!    of point 3 above), then w = 256 * m for some 0 <= m < 256.
+//! 5. Hence w / 256 ∈ {0, 1, ..., 255}
+//! 6. Hence y * 256 = L - w / 256 ∈ {-255, -254, ..., 2^16 - 1}
+//! 7. Taking the intersection of ranges in 2. and 6. we see that
+//!    y * 256 ∈ {0, 256, 512, ..., 2^16 - 256}
+//! 8. Hence y ∈ {0, 1, ..., 255}
+
+use ethereum_types::U256;
+use plonky2::field::extension::Extendable;
+use plonky2::field::packed::PackedField;
+use plonky2::field::types::{Field, PrimeField64};
+use plonky2::hash::hash_types::RichField;
+use plonky2::iop::ext_target::ExtensionTarget;
+use plonky2::plonk::circuit_builder::CircuitBuilder;
+use static_assertions::const_assert;
+
+use crate::arithmetic::columns::*;
+use crate::arithmetic::utils::u256_to_array;
+use crate::constraint_consumer::{ConstraintConsumer, RecursiveConstraintConsumer};
+
+// Give meaningful names to the columns of AUX_INPUT_REGISTER_0 that
+// we're using
+const BYTE_LAST_LIMB_LO: usize = AUX_INPUT_REGISTER_0.start + 7;
```

Comment:
Why start at `AUX_INPUT_REGISTER_0.start + 7` when `AUX_INPUT_REGISTER_0.start + 6` is unused?

Reply:
Because counting is haaaaard. :)